pennylane/math/quantum.py

```diff
@@ -628,3 +603,183 @@ def _compute_mutual_info(
+        return overlap
+    # First state mixed, second state pure
+    if state1.shape == (len_state1,) and state0.shape != (len_state0,):
+        overlap = np.tensordot(state0, np.transpose(np.conj(state1)), axes=1)
+        overlap = np.tensordot(state1, overlap, axes=1)
+        overlap = np.real(overlap)
+        return overlap
+    # First state pure, second state mixed
+    if state0.shape == (len_state0,) and state1.shape != (len_state1,):
+        overlap = np.tensordot(state1, np.transpose(np.conj(state0)), axes=1)
+        overlap = np.tensordot(state0, overlap, axes=1)
+        overlap = np.real(overlap)
+        return overlap
+    # Two mixed states
+    fid = _compute_fidelity(state0, state1)
+    return fid
+
+
+def sqrt_matrix(density_matrix):
+    r"""Compute the square root matrix of a density matrix where :math:`\rho = \sqrt{\rho} \times \sqrt{\rho}`
+    Args:
+        density_matrix (tensor_like): 2D density matrix of the quantum system.
+    Returns:
+        (tensor_like): Square root of the density matrix.
+    """
+    evs, vecs = qml.math.linalg.eigh(density_matrix)
+    evs = np.real(evs)
+    evs = qml.math.where(evs > 0.0, evs, 0.0)
+    if not is_abstract(evs):
+        evs = qml.math.cast_like(evs, vecs)
+    return vecs @ qml.math.diag(np.sqrt(evs)) @ np.conj(np.transpose(vecs))
+
+
+def _compute_fidelity(density_matrix0, density_matrix1):
+    r"""Compute the fidelity for two density matrices with the same number of wires.
+
+    .. math::
+            F( \rho , \sigma ) = -\text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+    """
+    # Implementation in single dispatches (sqrt(rho))
+    sqrt_mat = qml.math.sqrt_matrix(density_matrix0)
+
+    # sqrt(rho) * sigma * sqrt(rho)
+    sqrt_mat_sqrt = sqrt_mat @ density_matrix1 @ sqrt_mat
+
+    # extract eigenvalues
+    evs = qml.math.eigvalsh(sqrt_mat_sqrt)
+    evs = np.real(evs)
+    evs = qml.math.where(evs > 0.0, evs, 0.0)
+
+    trace = (qml.math.sum(qml.math.sqrt(evs))) ** 2
+
+    return trace
+
+
+def _check_density_matrix(density_matrix):
+    """Check the shape, the trace and the positive semi-definitiveness of a matrix."""
+    shape = density_matrix.shape[0]
+    if (
+        len(density_matrix.shape) != 2
+        or density_matrix.shape[0] != density_matrix.shape[1]
+        or not np.log2(shape).is_integer()
+    ):
+        raise ValueError("Density matrix must be of shape (2**N, 2**N).")
+    # Check trace
+    trace = np.trace(density_matrix)
+    if not is_abstract(trace):
+        if not allclose(trace, 1.0, atol=1e-10):
+            raise ValueError("The trace of the density matrix should be one.")
+        # Check if the matrix is Hermitian
+        conj_trans = np.transpose(np.conj(density_matrix))
+        if not allclose(density_matrix, conj_trans):
+            raise ValueError("The matrix is not Hermitian.")
+        # Check if positive semi-definite
+        evs = np.linalg.eigvalsh(density_matrix)
+        evs = np.real(evs)

```

Comment:
Side note (no action needed): we're computing eigenvalues quite a lot in this file, and it might be worth later on to store these values somehow instead of recomputing them everytime.

Reply:
Yes that would be a great improvement!